File: LAB2/MDAG.js
class BednarekTolbi {
    constructor(graph) {
        this.graph = graph;
        this.vertices = Object.keys(graph);
        this.n = this.vertices.length;

        this.stepLogs = [];
        this.S = [];
    }

    isIndependent(set) {
        for (let v of set) {
            for (let u of set) {
                if (v !== u && this.graph[v].includes(u)) {
                    return false;
                }
            }
        }
        return true;
    }

    unique(sets) {
        const seen = new Set();
        const res = [];

        for (const s of sets) {
            const key = [...new Set(s)].sort().join(",");
            if (!seen.has(key)) {
                seen.add(key);
                res.push([...s]);
            }
        }

        return res;
    }

    maximalSets(sets) {
        return sets.filter(a =>
            !sets.some(b =>
                b.length > a.length &&
                a.every(x => b.includes(x))
            )
        );
    }

    expand(family, v) {
        const keep = [];
        const expanded = [];

        for (const set of family) {
            keep.push([...set]);

            const cand = [...set, v];

            if (this.isIndependent(cand)) {
                expanded.push(cand);
            }
        }

        return { keep, expanded };
    }

    run() {
        let family = [[]];

        for (let i = 0; i < this.n; i++) {
            const v = this.vertices[i];

            const { keep, expanded } = this.expand(family, v);

            const merged = this.unique([...keep, ...expanded]);
            const maximal = this.maximalSets(merged);

            // сохраняем S_k
            this.S.push({
                k: i + 1,
                Sk: maximal
            });

            this.stepLogs.push({
                step: i + 1,
                vertex: v,
                previous: family,
                kept: keep,
                expanded: expanded,
                merged,
                maximal
            });

            family = maximal;
        }

        return family;
    }

    printSteps() {
        for (const log of this.stepLogs) {
            console.log("\n======================");
            console.log(`Шаг ${log.step} (добавляем ${log.vertex})`);

            console.log("Было:");
            console.table(log.previous);

            console.log("Без вершины:");
            console.table(log.kept);

            console.log("С вершиной:");
            console.table(log.expanded);

            console.log("После объединения:");
            console.table(log.merged);

            console.log("Максимальные:");
            console.table(log.maximal);
        }
    }

    printSk() {
        console.log("\nS_k (семейства максимальных независимых множеств)");

        for (const item of this.S) {
            console.log(`\nS_${item.k}:`);
            console.table(item.Sk);
        }
    }

    solve() {
        const result = this.run();

        const maxSize = Math.max(...result.map(s => s.length));

        return {
            maximal: result,
            largest: result.filter(s => s.length === maxSize),
            alpha: maxSize,
            S: this.S
        };
    }
}

const graph = {
    x1: ["x2", "x3"],
    x2: ["x1", "x4"],
    x3: ["x1", "x5"],
    x4: ["x2"],
    x5: ["x3"]
};

const solver = new BednarekTolbi(graph);

const result = solver.solve();

solver.printSteps();
solver.printSk();

console.log("\nМаксимальные:");
console.log(result.maximal);

console.log("\nНВУП:");
console.log(result.largest);

console.log("\nα(G):", result.alpha);

File: LAB2/MDAG.py
def find_max_independent_sets(vertices, edges):
    # Проверка: есть ли ребро между вершинами
    def is_adjacent(v1, v2):
        return (v1, v2) in edges or (v2, v1) in edges

    # Начинаем с первого элемента
    independent_sets = [{vertices[0]}]

    # Добавляем вершины по одной
    for i in range(1, len(vertices)):
        new_vertex = vertices[i]
        new_sets = []

        for s in independent_sets:
            # Проверяем: можно ли добавить новую вершину в множество
            if all(not is_adjacent(new_vertex, v) for v in s):
                # Если можно — добавляем
                new_sets.append(s | {new_vertex})
            else:
                # Если нельзя — оставляем как есть
                new_sets.append(s)

                # Пробуем создать новое множество с новой вершиной
                compatible = {v for v in s if not is_adjacent(v, new_vertex)}
                new_sets.append(compatible | {new_vertex})

        # Оставляем только максимальные множества
        maximal_sets = []
        for s in new_sets:
            if not any(s < other for other in new_sets):
                if s not in maximal_sets:
                    maximal_sets.append(s)

        independent_sets = maximal_sets

    return independent_sets


# --- ТАБЛИЦА СМЕЖНОСТИ (задаётся здесь) ---
table = [
    [0, 0, 1, 1, 1],  # x1
    [0, 0, 1, 0, 0],  # x2
    [1, 1, 0, 1, 0],  # x3
    [1, 0, 1, 0, 0],  # x4
    [1, 0, 0, 0, 0]  # x5
]


# --- ФОРМИРОВАНИЕ ВЕРШИН И РЁБЕР ПО ТАБЛИЦЕ ---
def formatTable():
    n = len(table)  # количество вершин
    vertices = [f"x{i + 1}" for i in range(n)]  # имена вершин: x1, x2, ...
    edges = []
    for i in range(n):
        for j in range(i + 1, n):  # проверяем только наддиагональные элементы
            if table[i][j] == 1:
                edges.append((vertices[i], vertices[j]))

    return vertices, edges


vertices, edges = formatTable()
print("Вершины:", vertices)
print("Рёбра:", edges)

# --- ЗАПУСК И ВЫВОД РЕЗУЛЬТАТА ---
result = find_max_independent_sets(vertices, edges)

print("\nМаксимальные независимые множества:")
for i, s in enumerate(result, 1):
    print(f"M{i}: {sorted(s)}")

# Наибольшее множество
max_size = max(len(s) for s in result)
largest = [s for s in result if len(s) == max_size]

print("\nРазмер α₀ =", max_size)
print("Наибольшие внутреустойчивые подмножества:", [sorted(s) for s in largest])
